File: datasets/C++/Deepseek/Algorithm/medium4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <sstream>

using namespace std;

vector<string> grid;
vector<string> word_list;
vector<tuple<char, int, int, int>> slots;

bool backtrack(int word_index) {
    if (word_index == word_list.size()) {
        return true;
    }
    
    string current_word = word_list[word_index];
    
    for (auto& slot : slots) {
        char direction = get<0>(slot);
        int row = get<1>(slot);
        int col = get<2>(slot);
        int length = get<3>(slot);
        
        if (length != current_word.length()) {
            continue;
        }
        
        bool can_place = true;
        vector<char> original_chars;
        
        if (direction == 'h') {
            for (int k = 0; k < length; k++) {
                char cell = grid[row][col + k];
                if (cell != '-' && cell != current_word[k]) {
                    can_place = false;
                    break;
                }
                original_chars.push_back(cell);
            }
        } else { // vertical
            for (int k = 0; k < length; k++) {
                char cell = grid[row + k][col];
                if (cell != '-' && cell != current_word[k]) {
                    can_place = false;
                    break;
                }
                original_chars.push_back(cell);
            }
        }
        
        if (can_place) {
            // Place the word
            if (direction == 'h') {
                for (int k = 0; k < length; k++) {
                    grid[row][col + k] = current_word[k];
                }
            } else {
                for (int k = 0; k < length; k++) {
                    grid[row + k][col] = current_word[k];
                }
            }
            
            // Recurse
            if (backtrack(word_index + 1)) {
                return true;
            }
            
            // Backtrack
            if (direction == 'h') {
                for (int k = 0; k < length; k++) {
                    grid[row][col + k] = original_chars[k];
                }
            } else {
                for (int k = 0; k < length; k++) {
                    grid[row + k][col] = original_chars[k];
                }
            }
        }
    }
    
    return false;
}

vector<string> crosswordPuzzle(vector<string> crossword, string words) {
    grid = crossword;
    
    // Split words by semicolon
    stringstream ss(words);
    string word;
    while (getline(ss, word, ';')) {
        word_list.push_back(word);
    }
    
    // Find all horizontal and vertical slots
    // Horizontal slots
    for (int i = 0; i < 10; i++) {
        int j = 0;
        while (j < 10) {
            if (grid[i][j] == '-') {
                int start = j;
                while (j < 10 && grid[i][j] == '-') {
                    j++;
                }
                int length = j - start;
                if (length > 1) {
                    slots.emplace_back('h', i, start, length);
                }
            } else {
                j++;
            }
        }
    }
    
    // Vertical slots
    for (int j = 0; j < 10; j++) {
        int i = 0;
        while (i < 10) {
            if (grid[i][j] == '-') {
                int start = i;
                while (i < 10 && grid[i][j] == '-') {
                    i++;
                }
                int length = i - start;
                if (length > 1) {
                    slots.emplace_back('v', start, j, length);
                }
            } else {
                i++;
            }
        }
    }
    
    // Sort slots by length (descending)
    sort(slots.begin(), slots.end(), 
        [](const auto& a, const auto& b) {
            return get<3>(a) > get<3>(b);
        });
    
    // Start backtracking
    backtrack(0);
    
    return grid;
}

int main() {
    vector<string> crossword(10);
    for (int i = 0; i < 10; i++) {
        cin >> crossword[i];
    }
    
    string words;
    cin >> words;
    
    vector<string> result = crosswordPuzzle(crossword, words);
    for (const string& row : result) {
        cout << row << endl;
    }
    
    return 0;
}

File: datasets/Python/Deepseek/Algorithm/medium4.py
def crosswordPuzzle(crossword, words):
    # Convert the crossword to a list of lists for easier manipulation
    grid = [list(row) for row in crossword]
    word_list = words.split(';')
    
    # Find all horizontal and vertical slots
    slots = []
    
    # Find horizontal slots (left to right)
    for i in range(10):
        j = 0
        while j < 10:
            if grid[i][j] == '-':
                start = j
                while j < 10 and grid[i][j] == '-':
                    j += 1
                length = j - start
                if length > 1:
                    slots.append(('h', i, start, length))
            else:
                j += 1
    
    # Find vertical slots (top to bottom)
    for j in range(10):
        i = 0
        while i < 10:
            if grid[i][j] == '-':
                start = i
                while i < 10 and grid[i][j] == '-':
                    i += 1
                length = i - start
                if length > 1:
                    slots.append(('v', start, j, length))
            else:
                i += 1
    
    # Sort slots by length (descending) to try longer words first
    slots.sort(key=lambda x: -x[3])
    
    # Backtracking function
    def backtrack(word_index):
        if word_index == len(word_list):
            return True
        
        current_word = word_list[word_index]
        
        for slot in slots:
            direction, row, col, length = slot
            if length != len(current_word):
                continue
            
            # Check if this word can fit in this slot
            can_place = True
            original_chars = []
            
            if direction == 'h':
                for k in range(length):
                    cell = grid[row][col + k]
                    if cell != '-' and cell != current_word[k]:
                        can_place = False
                        break
                    original_chars.append(cell)
            else:  # vertical
                for k in range(length):
                    cell = grid[row + k][col]
                    if cell != '-' and cell != current_word[k]:
                        can_place = False
                        break
                    original_chars.append(cell)
            
            if can_place:
                # Place the word
                if direction == 'h':
                    for k in range(length):
                        grid[row][col + k] = current_word[k]
                else:
                    for k in range(length):
                        grid[row + k][col] = current_word[k]
                
                # Recurse
                if backtrack(word_index + 1):
                    return True
                
                # Backtrack
                if direction == 'h':
                    for k in range(length):
                        grid[row][col + k] = original_chars[k]
                else:
                    for k in range(length):
                        grid[row + k][col] = original_chars[k]
        
        return False
    
    # Start backtracking
    backtrack(0)
    
    # Convert the grid back to strings
    return [''.join(row) for row in grid]

# Read input
crossword = [input().strip() for _ in range(10)]
words = input().strip()

# Solve and print
result = crosswordPuzzle(crossword, words)
for row in result:
    print(row)